src/goto-instrument/synthesizer/expr_enumerator.cpp

```diff
@@ -0,0 +1,375 @@
+// (ones[0], ones[1]-ones[0], ..., ones[k-2]-ones[k-3], n-ones[k-2]).
+partitiont from_bits_to_partition(std::size_t v, std::size_t n)
+{
+  const std::vector<std::size_t> ones_pos = get_ones_pos(v);
+
+  INVARIANT(ones_pos.size() >= 1, "There should be at least one bit set in v");
+
+  partitiont result = {ones_pos[0]};
+
+  for(std::size_t i = 1; i < ones_pos.size(); i++)
+  {
+    result.emplace_back(ones_pos[i] - ones_pos[i - 1]);
+  }
+  result.emplace_back(n - ones_pos[ones_pos.size() - 1]);
+
+  return result;
+}
+
+std::list<partitiont> non_leaf_enumeratort::get_partitions(
+  const std::size_t n,
+  const std::size_t k) const
+{
+  // Every component should contain at least one element.
+  if(n < k)
+    return {};
+
+  // This new implementation works only for n no larger than 64.
+  if(n > 64)
+    return get_partitions_64(n, k);
+
+  // Number of bits at all.
+  const std::size_t length = sizeof(std::size_t) * 8;
+
+  // We enumerate all bit vectors `v` with k-1 one's such that each component
+  // corresponds to one unique partition.
+  // For a bit vector with ones at positions (computed by `get_ones_pos`)
+  // (ones[0], ones[1], ..., ones[k-2]),
+  // the corresponding partition is
+  // (ones[0], ones[1]-ones[0], ..., ones[k-2]-ones[k-3], n-ones[k-2]).
+
+  // Initial `v` is with ones at positions (n-k+1, n-k+2, ..., n-2, n-1).
+  std::size_t v = 0;
+  // Initial `end` (the last bit vectorr we enumerate) is with ones at
+  // positions (1, 2, 3, ..., k-1).
+  std::size_t end = 0;
+  for(size_t i = 0; i < k - 1; i++)
+  {
+    v++;
+    v = v << 1;
+    end++;
+    end = end << 1;
+  }
+  v = v << (length - n);
+  end = end << (length - k);
+
+  std::list<partitiont> result;
+  while(v != end)
+  {
+    // Construct the parititon for current bit vector and add it to `result`
+    result.emplace_back(from_bits_to_partition(v, n));
+
+    // https://graphics.stanford.edu/~seander/bithacks.html#NextBitPermutation
+    // Compute the lexicographically next bit permutation.
+    std::size_t t = (v | (v - 1)) + 1;
+    v = t | ((((t & -t) / (v & -v)) >> 1) - 1);
+  }
+  result.emplace_back(from_bits_to_partition(v, n));
+
+  return result;
+}
+
+bool binary_functional_enumeratort::is_commutative(const irep_idt &op) const
+{
+  return op_id == ID_equal || op_id == ID_plus || op_id == ID_notequal ||
+         op_id == ID_or || op_id == ID_and || op_id == ID_xor ||
+         op_id == ID_bitand || op_id == ID_bitor || op_id == ID_bitxor ||
+         op_id == ID_mult;
+}
+
+bool binary_functional_enumeratort::is_equivalence_class_representation(
+  const expr_listt &exprs) const
+{
+  // When the two sub-enumerators are exchangeable---they enumerate the same
+  // set of expressions---, and the operator is commutative, `exprs` is a
+  // representation if its sub-expressions are sorted.
+  if(
+    is_exchangeable && is_commutative(op_id) &&
+    from_expr(exprs.front()) > from_expr(exprs.back()))
+    return false;
```

Comment:
Nit pick: braces around body when the condition spans multiple lines.

Reply:
It is required by check-clang-format.

Reply:
Apologies for being unclear: yes, up to line 300 this is fine, but then do
```
{
  return false;
}
```